### PR TITLE
[Analyzer] ip2location.io 

### LINKED ISSUE #1924: [Analyzer] IP2Location.io
## Name
IP2Location.io

## Link
https://www.ip2location.io

## Type of analyzer
Observable -> IP address only

## Why should we use it
It detects IP geolocation and proxy information using API. Free to use.

## Possible implementation
Refer to the documentation section in API page.

### PR DESCRIPTION
**Pull request by aryan-bhokare**
closes #1924

# Description
Added new analyzer that provides a fast and accurate IP Geolocation API tool to determine a user's location and use the geolocation information in different use cases.

## Type of change
- [x] New feature (non-breaking change which adds functionality).

# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [x] A new plugin (analyzer, connector, visualizer, playbook, pivot or ingestor) was added or changed, in which case:
    - [x] I strictly followed the documentation ["How to create a Plugin"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-add-a-new-plugin)
    - [x] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [] [Advanced-Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [x] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [x] I have dumped the configuration from Django Admin using the `dumpplugin` command and added it in the project as a data migration. (["How to share a plugin with the community"](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-share-your-plugin-with-the-community))
    - [] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [x] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook by following [this guide](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-modify-a-plugin).
    - [] Check if it could make sense to add that analyzer/connector to other [freely available playbooks](https://intelowl.readthedocs.io/en/develop/Usage.html#list-of-pre-built-playbooks).
    - [] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Contribute.md)).
- [] If the GUI has been modified:
    - [] I have a provided a screenshot of the result in the PR.
    - [] I have created new frontend tests for the new component or updated existing ones.

### Important Rules
- If you miss to compile the Checklist properly, your PR won't be reviewed by the maintainers.
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review. After being reviewed and received a "change request", you should explicitly ask for a review again once you have made the requested changes.

### CODE REVIEW

**Comment by aryan-bhokare** [comment]
@mlodic can you take a look.

File: api_app/analyzers_manager/migrations/0058_analyzer_config_ip2location.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    ManyToManyDescriptor,
+)
+
+plugin = {
+    "name": "ip2location",
+    "python_module": {
+        "module": "ip2location.Ip2location",
+        "base_path": "api_app.analyzers_manager.observable_analyzers",
+    },
+    "description": "IP2Location.io allows users to check IP address location in real time.",
```
Comment (mlodic) [inline api_app/analyzers_manager/migrations/0058_analyzer_config_ip2location.py:14]: can you please add a link to the service main page?

File: api_app/analyzers_manager/migrations/0058_analyzer_config_ip2location.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+params = [
+    {
+        "python_module": {
+            "module": "ip2location.Ip2location",
+            "base_path": "api_app.analyzers_manager.observable_analyzers",
+        },
+        "name": "_api_key_name",
+        "type": "str",
+        "description": "",
+        "is_secret": False,
```
Comment (mlodic) [inline api_app/analyzers_manager/migrations/0058_analyzer_config_ip2location.py:40]: this should be True

File: api_app/analyzers_manager/observable_analyzers/ip2location.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    def get_response(self, payload):
+        return requests.get(self.base_url, params=payload)
+
+    def run(self):
+        try:
+            payload = {"ip": self.observable_name}
+
+            # Add key to payload if the API version is "keyed"
+            if self.ip2location_api_version == "keyed":
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/ip2location.py:23]: you need to define this parameter as an attribute of the class. Please explain the difference between the keyless and the keyed option

File: api_app/analyzers_manager/observable_analyzers/ip2location.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                payload["key"] = self._api_key_name
+
+            location_info = self.get_response(payload)
+            location_info.raise_for_status()
+
+        except requests.RequestException as e:
+            raise AnalyzerRunException(e)
+
+        response = {"location_info": location_info.json()}
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/ip2location.py:32]: no need to create a wrapper dictionary. Just use the response dict itself

File: api_app/analyzers_manager/observable_analyzers/ip2location.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        return response
+
+    @classmethod
+    def _monkeypatch(cls):
+        patches = [
+            if_mock_connections(
+                patch(
+                    "requests.get",
+                    return_value=MockUpResponse({}, 200),
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/ip2location.py:41]: please add a sample of a possible response from the service here

**Review by mlodic — CHANGES_REQUESTED** [review]
please check the checklist

File: api_app/analyzers_manager/migrations/0058_analyzer_config_ip2location.py (hunk 1 above)
Comment (aryan-bhokare) [reply api_app/analyzers_manager/migrations/0058_analyzer_config_ip2location.py]: Added

File: api_app/analyzers_manager/migrations/0058_analyzer_config_ip2location.py (hunk 2 above)
Comment (aryan-bhokare) [reply api_app/analyzers_manager/migrations/0058_analyzer_config_ip2location.py]: changed it to true

File: api_app/analyzers_manager/observable_analyzers/ip2location.py (hunk 3 above)
Comment (aryan-bhokare) [reply api_app/analyzers_manager/observable_analyzers/ip2location.py:27]: done please take a look

File: api_app/analyzers_manager/observable_analyzers/ip2location.py (hunk 4 above)
Comment (aryan-bhokare) [reply api_app/analyzers_manager/observable_analyzers/ip2location.py]: my bad changed it.

File: api_app/analyzers_manager/observable_analyzers/ip2location.py (hunk 5 above)
Comment (aryan-bhokare) [reply api_app/analyzers_manager/observable_analyzers/ip2location.py]: done!

File: api_app/analyzers_manager/observable_analyzers/ip2location.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+from api_app.analyzers_manager import classes
+from api_app.analyzers_manager.exceptions import AnalyzerRunException
+from tests.mock_utils import MockUpResponse, if_mock_connections, patch
+
+
+class Ip2location(classes.ObservableAnalyzer):
+    base_url: str = "https://api.ip2location.io/"
+    ip2location_api_key: str
+    ip2location_api_version: str
```
Comment (mlodic) [inline api_app/analyzers_manager/observable_analyzers/ip2location.py:14]: there is no need to add ip2location at the start of these attributes, plus the secret "api_key" must be preprended by `_` as stated in the documentation.

**Review by mlodic — CHANGES_REQUESTED** [review]
please check the checklist again as I have already asked, it is really important to follow it strictly.

File: api_app/analyzers_manager/observable_analyzers/ip2location.py (hunk 6 above)
Comment (aryan-bhokare) [reply api_app/analyzers_manager/observable_analyzers/ip2location.py:14]: I am having one issue there. Even though have the _api_key_name value present in the parameter still it's not working.
WhatsApp Image 2024-01-22 at 11 56 30 AM
image

Comment (mlodic) [reply api_app/analyzers_manager/observable_analyzers/ip2location.py:14]: what does that error shown up in the IntelOwl report means to you?

Comment (aryan-bhokare) [reply api_app/analyzers_manager/observable_analyzers/ip2location.py:14]: I changed the variable name locally to _api_key_name.
Even when I did not changed it it was showing ip2location_api_key.
My apologies for not mentioning about the change :)

Comment (aryan-bhokare) [reply api_app/analyzers_manager/observable_analyzers/ip2location.py:14]: somehow it's not getting the value of _api_key_name from the plugin config, Do we need to initalise it something?

**Comment by aryan-bhokare** [comment]
I will be closing this pr and opening a new one.

**Pull request closed without merging**